### PR TITLE
BUG: DTI/TDI.insert doing invalid casting

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [ ] closes #xxxx
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

File: pandas/tests/indexes/timedeltas/test_insert.py [hunk 1]
```diff
@@ -82,6 +82,17 @@ def test_insert_invalid_na(self):
...
+    @pytest.mark.parametrize(
+        "value", [0, np.int64(0), np.float64(0), np.array(0), np.datetime64(456, "us")]
+    )
+    def test_insert_mismatched_types(self, value):
+        # GH#33703 dont cast these to td64
+        tdi = TimedeltaIndex(["4day", "1day", "2day"], name="idx")
+
+        msg = "incompatible label"
+        with pytest.raises(TypeError, match=msg):
+            tdi.insert(1, value)
```
Comment (simonjayhawkins) [inline pandas/tests/indexes/timedeltas/test_insert.py:94]: the second parameter is named `item`. my preference when paramatrising is to use the same name.

File: pandas/tests/indexes/timedeltas/test_insert.py [hunk 2]
```diff
@@ -82,6 +82,17 @@ def test_insert_invalid_na(self):
         with pytest.raises(TypeError, match="incompatible label"):
             idx.insert(0, np.datetime64("NaT"))
 
+    @pytest.mark.parametrize(
+        "value", [0, np.int64(0), np.float64(0), np.array(0), np.datetime64(456, "us")]
+    )
+    def test_insert_mismatched_types(self, value):
```
Comment (simonjayhawkins) [inline pandas/tests/indexes/timedeltas/test_insert.py:88]: unwritten convention

```suggestion
    def test_insert_mismatched_types_raises(self, value):
```

File: pandas/tests/indexes/datetimes/test_insert.py [hunk 3]
```diff
@@ -165,3 +165,26 @@ def test_insert(self):
                 assert result.name == expected.name
                 assert result.tz == expected.tz
                 assert result.freq is None
+
+    @pytest.mark.parametrize(
+        "value", [0, np.int64(0), np.float64(0), np.array(0), np.timedelta64(456)]
+    )
+    def test_insert_mismatched_types(self, tz_aware_fixture, value):
```
Comment (simonjayhawkins) [inline pandas/tests/indexes/datetimes/test_insert.py:172]:
```suggestion
    def test_insert_mismatched_types_raises(self, tz_aware_fixture, value):
```

**Review by simonjayhawkins — COMMENTED** [review]
does this closes any issues?

File: pandas/tests/indexes/timedeltas/test_insert.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/tests/indexes/timedeltas/test_insert.py:94]: good idea

**Comment by jbrockmendel** [comment]
> does this closes any issues?

Not AFAICT

**Comment by jreback** [comment]
looks good, can you merge master and ping on green.

**Comment by jbrockmendel** [comment]
rebased+green

**Pull request merged**